beacon-chain/casper/state_transition.go

```diff
@@ -0,0 +1,95 @@
+package casper
+
+import (
+	"github.com/prysmaticlabs/prysm/beacon-chain/params"
+	"github.com/prysmaticlabs/prysm/beacon-chain/utils"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/bitutil"
+)
+
+func TallyVoteBalances(
+	blockHash [32]byte,
+	slot uint64,
+	blockVoteCache map[[32]byte]*utils.VoteCache,
+	validators []*pb.ValidatorRecord,
+	timeSinceFinality uint64,
+	enableRewardChecking bool) (uint64, []*pb.ValidatorRecord) {
+	var blockVoteBalance uint64
+
+	if _, ok := blockVoteCache[blockHash]; ok {
+		blockVoteBalance = blockVoteCache[blockHash].VoteTotalDeposit
+		voterIndices := blockVoteCache[blockHash].VoterIndices
+
+		// Apply Rewards for each slot.
+		if enableRewardChecking {
+			validators = CalculateRewards(
+				slot,
+				voterIndices,
+				validators,
+				blockVoteBalance,
+				timeSinceFinality)
+		}
+	} else {
+		blockVoteBalance = 0
+	}
+
+	return blockVoteBalance, validators
+}
+
+func FinalizeAndJustifySlots(
+	slot uint64, justifiedSlot uint64, finalizedSlot uint64,
+	justifiedStreak uint64, blockVoteBalance uint64, totalDeposits uint64) (uint64, uint64, uint64) {
+
+	if 3*blockVoteBalance >= 2*totalDeposits {
+		if slot > justifiedSlot {
+			justifiedSlot = slot
+		}
+		justifiedStreak++
+	} else {
+		justifiedStreak = 0
+	}
+
+	if slot > params.GetConfig().CycleLength && justifiedStreak >= params.GetConfig().CycleLength+1 && slot-params.GetConfig().CycleLength-1 > finalizedSlot {
```

Comment:
this is a super long conditional - consider splitting up into boolean variables that have comments in order to understand what's going on